### PR TITLE
[WebPubSub] Implement client sdk lifetime

### PR DESCRIPTION
**Pull request by zackliu**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 1]
```diff
@@ -328,24 +587,397 @@ internal async Task<WebPubSubResult> SendMessageWithAckAsync(Func<ulong, WebPubS
...
+            {
+                await (ServerMessageReceived?.Invoke(eventArgs) ?? Task.CompletedTask).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                WebPubSubClientEventSource.Log.FailedToInvokeEvent(nameof(ServerMessageReceived), ex.Message);
+            }
+        }
+
+        private async Task SafeInvkeConnectedAsync(WebPubSubConnectedEventArgs eventArgs)
```
Comment (JialinXin) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:657]: Typo: Invke => Invoke. And a few below.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 2]
```diff
@@ -28,24 +28,60 @@ namespace Azure.Messaging.WebPubSub.Clients
...
         private const string ErrorNameDuplicate = "Duplicate";
+        private const string RecoverConnectionIdQuery = "awps_connection_id";
+        private const string RecoverReconnectionTokenQuery = "awps_reconnection_token";
 
         // Some exposed properties for testing
         internal IWebSocketClientFactory WebSocketClientFactory { get; set; }
         internal TimeSpan RecoverDelay { get; set; } = TimeSpan.FromSeconds(1);
+        internal ulong CurrentSequenceId => _sequenceId.Current;
+
+        private static readonly UnboundedChannelOptions s_unboundedChannelOptions = new UnboundedChannelOptions
```
Comment (vwxyzh) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:39]: i dont think we need cache it. each client only creates two option instances.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 3]
```diff
@@ -28,24 +28,60 @@ namespace Azure.Messaging.WebPubSub.Clients
     public class WebPubSubClient
     {
         private const string ErrorNameDuplicate = "Duplicate";
+        private const string RecoverConnectionIdQuery = "awps_connection_id";
+        private const string RecoverReconnectionTokenQuery = "awps_reconnection_token";
 
         // Some exposed properties for testing
         internal IWebSocketClientFactory WebSocketClientFactory { get; set; }
         internal TimeSpan RecoverDelay { get; set; } = TimeSpan.FromSeconds(1);
+        internal ulong CurrentSequenceId => _sequenceId.Current;
```
Comment (vwxyzh) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:37]: order, `const`, `static readonly`, `static`, `readonly`, others.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs (hunk 2 above)
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:39]: I don't mean to cache it, but it makes codes look clean

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs (hunk 3 above)
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs]: I moved to property part

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 4]
```diff
@@ -328,24 +587,397 @@ internal async Task<WebPubSubResult> SendMessageWithAckAsync(Func<ulong, WebPubS
...
+        {
+            foreach (var entity in _ackCache)
+            {
+                if (_ackCache.TryRemove(entity.Key, out var value))
+                {
+                    value.SetException(new SendMessageFailedException("Connection is disconnected before receive ack from the service", entity.Value.AckId));
+                }
+            }
+
+            if (closeStatus == WebSocketCloseStatus.PolicyViolation)
```
[the remark this answers is not in the document]
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:756]: It's designed by us. The service side use `PolicyViolation` code to indicate unrecoverable.

**Review by JialinXin — APPROVED** [review]

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 5]
```diff
@@ -114,7 +235,48 @@ public virtual Task StartAsync(CancellationToken cancellationToken = default)
         public virtual Task StopAsync()
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         {
-            throw new NotImplementedException();
+            ThrowIfDisposed();
+
+            lock (_stopLock)
```
Comment (vicancy) [inline sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:240]: it's a deadlock when :
```
StopAsync(); // entering the _stopLock waiting for _connectionLock
StartAsync(); // entering the _connectionLock waiting for _stopLock
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 6]
```diff
@@ -415,6 +1047,55 @@ public AckEntity(ulong ackId)
...
+            private ulong _sequenceId;
+            private bool _updated;
+
+            public ulong Current => _sequenceId;
+
+            public bool TryUpdate(ulong sequenceId)
+            {
+                lock (_lock)
+                {
+                    _updated = true;
```
[the remark this answers is not in the document]
Comment (vicancy) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:1062]: So this only happens in recover case? Please add some comment to this implementation.

File: sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs [hunk 7]
```diff
@@ -291,9 +468,91 @@ internal virtual async Task<WebPubSubResult> SendEventAttemptAsync(string eventN
...
+                    if (result.Payload.Length > 0)
+                    {
+                        try
+                        {
+                            var message = _protocol.ParseMessage(result.Payload);
+                            await HandleMessageAsync(message, token).ConfigureAwait(false);
+                        }
+                        catch (Exception ex)
+                        {
+                            WebPubSubClientEventSource.Log.FailedToProcessMessage(ex.Message);
```
[the remark this answers is not in the document]
Comment (zackliu) [reply sdk/webpubsub/Azure.Messaging.WebPubSub.Client/src/WebPubSubClient.cs:526]: I don't think so. It shouldn't be an error that needs to close the connection.

**Review by vicancy — APPROVED** [review]

**Pull request merged**